src/featbin/wav-reverberate.cc

```diff
@@ -0,0 +1,331 @@
+               << " num_samp " << num_samp_rir
+               << " num_chan " << num_chan_rir;
+    if (!multi_channel_output) {
+      KALDI_ASSERT(rir_channel < num_chan_rir);
+    }
+
+    Matrix<BaseFloat> noise_matrix;
+    if (!noise_file.empty()) {
+      WaveData noise_wave;
+      {
+        Input ki(noise_file);
+        noise_wave.Read(ki.Stream());
+      }
+      noise_matrix = noise_wave.Data();
+      BaseFloat samp_freq_noise = noise_wave.SampFreq();
+      int32 num_samp_noise = noise_matrix.NumCols(),
+            num_chan_noise = noise_matrix.NumRows();
+      KALDI_LOG << "samp_freq of noise is " << samp_freq_noise
+                << " num_samp " << num_samp_noise
+                << " num_chan " << num_chan_noise;
+      if (multi_channel_output) {
+        KALDI_ASSERT(num_chan_rir == num_chan_noise);
+      } else {
+        KALDI_ASSERT(noise_channel < num_chan_noise);
+      }
+    }
+
+    int out_chan = 0;
+    int rir_chan = rir_channel;
+    int noise_chan = noise_channel;
+    int num_chan_output = 1;
+    if (multi_channel_output) {
+      num_chan_output = num_chan_rir;
+      rir_chan = 0;
+      noise_chan = 0;
+    }
+
+    Matrix<BaseFloat> out_matrix(num_chan_output, num_samp_input);
+
+    do {
+      Vector<BaseFloat> input(num_samp_input);
+      input.CopyRowFromMat(input_matrix, input_wave_channel);
+      Vector<BaseFloat> rir(num_samp_rir);
+      rir.CopyRowFromMat(rir_matrix, rir_chan);
+      rir.Scale(1 / static_cast<float>(2147483647));  // Assume 32 bits 
+
+      Vector<BaseFloat> noise(0);
+      if (!noise_file.empty()) {
+        float input_power = ComputeEarlyReverbEnergy(rir, input, samp_freq_rir);
+
+        noise.Resize(noise_matrix.NumCols());
+        noise.CopyRowFromMat(noise_matrix, noise_chan);
+        float noise_power = VecVec(noise, noise) / noise.Dim();
+
+        KALDI_VLOG(1) << "input_power of channel " << out_chan << " is " << input_power;
+        KALDI_VLOG(1) << "noise_power of channel " << out_chan << " is " << noise_power;
+
+        float scale_factor = sqrt(pow(10, -snr_db / 10) * input_power / noise_power);
+        noise.Scale(scale_factor);
+        KALDI_VLOG(1) << "scale_factor of channel " << out_chan << " : " << scale_factor << " for SNR db " << snr_db;
+      }
+
+      if (test_mode) {
+        Vector<BaseFloat> input_test(input);
+        ConvolveSignals(rir, &input_test);
+        BlockConvolveSignals(rir, &input);
+        AssertEqual(input, input_test, 0.001*input.Dim());
+      } else {
+        BlockConvolveSignals(rir, &input);
+      }
+
+      if (noise.Dim() > 0) {
+        AddVectorsOfUnequalLength(noise, &input);
+      }
+
+      float input_max = MaxAbsolute(input);
+
+      for (int32 j = 0; j < input.Dim(); j++)
+        input(j) = (0.99 * input(j) / input_max) * pow(2, 15);
+
+      out_matrix.CopyRowFromVec(input, out_chan);
+      out_chan++;
+      rir_chan++;
+      noise_chan++;
+    } while (out_chan < num_chan_output);
+
```

Comment:
Above, the use of a do... while loop makes the code hard to read, as do the variables out_chan, rir_chan and noise_chan.
What would be clearer, I think, is something like this:
Firstly, if multi_channel_output is true, then check that the various other channel options were not set, and print an error message and die if so.
Then do something like

int32 num_output_channels = (multi_channel_output ? num_chan_rir : 1);
  for (int32 output_channel = 0; output_channel < num_output_channels;
       output_channel++) {
    int32 this_rir_channel = (multi_channel_output ? output_channel : rir_channel);
   // it's not good to use differently-spelled versions of the same thing as distinct variables, e.g. chan versus channel.